text/0000-extern-impl.md

```diff
@@ -0,0 +1,624 @@
+dependant.
+
+We extend the `rustc` `--extern` flag with an additional `impl` flag: `--extern
+impl:dbtrait=dbtrait.rlib`. This means that the crate currently being compiled
+with this option, is allowed to implement traits and methods for types defined
+in the `dbtrait` crate (the *defining* crate).
+
+For example:
+```rust
+//! Defining crate
+trait DBTrait {}
+```
+```rust
+//! Implementation crate, also depends on the `mysql` crate.
+use mysql::MySql;
+
+// Allowed because we're depending on `dbtrait` with `impl`
+impl dbtrait::DBTrait for MySql {}
+```
+
+In other words, the implementing crate is considered the same as the defining
+crate as far as the orphan rule goes. Note that you may only implement traits
+and types for crates directly defined in the defining crate &mdash; you cannot
+implement re-exported types.
+
+The defining and implementing crates are not the same in other respects. They
+still have distinct crate names, defining their own crate-level namespaces. For
+example, if they both defined a type `Foo`, it would be two definitions
+`defining::Foo` and `implementing::Foo`.
+
+There are no other constraints on the defining and implementing crates - they
+can freely define new traits, types and other items, and implementations for
+those types and traits. An implementing crate can be a defining crate for
+another crate.
+
+Regardless, the definitions within the defining crate, the implementing crate,
+and between the implementing crate and defining crate must meet the current
+coherence requirements. Also if there are multiple implementation crates in
+scope in the dependency graph, they must also be coherent.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+[Coherence](https://rust-lang.github.io/chalk/book/clauses/coherence.html) is
+currently defined in terms of "local" definitions - that is, items defined with
+in the current crate.
+
+The details of the coherence checking logic are unchanged; the only change is
+the constraints on which crates a definition and an implementation can be in.
+
+The discussion below talks about coherence in terms of the crate dependency
+relations, but the actual coherence check is at the level of individual type and
+trait implementations. The crate dependency relations set an upper bound on
+this, as any crate which does not appear within the relevent dependency graph
+need not be considered for coherence checking.
+
+## Changes to coherence checking
+
+The key problem we need to solve is how to reconcile the Rust language-level
+details (types and traits) with the build system view (dependency relationship
+between crates). Coherence is defined in terms of type/trait definitions and
+implementations, which forms its own graph. This graph is embedded in the crate
+dependency graph &mdash; that is there are no edges in the type/trait
+implementation graph which are not also present in the crate dependency graph.
```

Comment:
Either I'm misunderstanding the point, or this is false for blanket impls

Reply:
Yes, good call. This needs to be worded more carefully.